common/src/main/java/com/aliyun/emr/rss/common/utils/ShutdownHookManager.java

```diff
@@ -0,0 +1,330 @@
+ * `rss.shutdown.timeout`.
+ *
+ * Note: code refer to Hadoop's ShutdownHookManager.
+ */
+public final class ShutdownHookManager {
+
+  private static final ShutdownHookManager MGR = new ShutdownHookManager();
+
+  private static final Logger LOG = LoggerFactory.getLogger(ShutdownHookManager.class);
+
+  /** Minimum shutdown timeout: {@value} second(s). */
+  public static final long TIMEOUT_MINIMUM = 1;
+
+  /** The default time unit used: seconds. */
+  public static final TimeUnit TIME_UNIT_DEFAULT = TimeUnit.SECONDS;
+
+  private static final ExecutorService EXECUTOR =
+      Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
+          .setDaemon(true)
+          .setNameFormat("shutdown-hook-%01d")
+          .build());
+
+  static {
+    try {
+      Runtime.getRuntime().addShutdownHook(
+          new Thread() {
+            @Override
+            public void run() {
+              if (MGR.shutdownInProgress.getAndSet(true)) {
+                LOG.info("Shutdown process invoked a second time: ignoring");
+                return;
+              }
+              long started = System.currentTimeMillis();
+              int timeoutCount = executeShutdown();
+              long ended = System.currentTimeMillis();
+              LOG.debug(String.format(
+                  "Completed shutdown in %.3f seconds; Timeouts: %d",
+                  (ended-started)/1000.0, timeoutCount));
+              // each of the hooks have executed; now shut down the
+              // executor itself.
+              shutdownExecutor(new RssConf());
+            }
+          }
+      );
+    } catch (IllegalStateException ex) {
+      // JVM is being shut down. Ignore
+      LOG.warn("Failed to add the ShutdownHook", ex);
+    }
+  }
+
+  /**
+   * Execute the shutdown.
+   * This is exposed purely for testing: do not invoke it.
+   * @return the number of shutdown hooks which timed out.
+   */
+  static int executeShutdown() {
+    int timeouts = 0;
+    for (HookEntry entry: MGR.getShutdownHooksInOrder()) {
+      Future<?> future = EXECUTOR.submit(entry.getHook());
+      try {
+        future.get(entry.getTimeout(), entry.getTimeUnit());
+      } catch (TimeoutException ex) {
+        timeouts++;
+        future.cancel(true);
+        LOG.warn("ShutdownHook '" + entry.getHook().getClass().getSimpleName() + "' timeout, "
+            + ex.toString(), ex);
+      } catch (Throwable ex) {
+        LOG.warn("ShutdownHook '" + entry.getHook().getClass().getSimpleName() + "' failed, "
+            + ex.toString(), ex);
+      }
+    }
+    return timeouts;
+  }
+
+  /**
+   * Shutdown the executor thread itself.
+   * @param conf the configuration containing the shutdown timeout setting.
+   */
+  private static void shutdownExecutor(final RssConf conf) {
+    try {
+      EXECUTOR.shutdown();
+      long shutdownTimeout = getShutdownTimeout(conf);
+      if (!EXECUTOR.awaitTermination(
+          shutdownTimeout,
+          TIME_UNIT_DEFAULT)) {
```

Comment:
getShutdownTimeout returns time in milliseconds, but here use default of seconds, should make them consistent.

Reply:
Done